File: react/src/js/http/ContactApi.ts
import {HttpService} from "../common/services/HttpService";
import {CreateParentChildrenRequest, Concession, ConcessionType, ContactFields, ContactFieldsRequest,
  SubmitFieldsRequest, Contact, ContactId, CreateContactParams} from "../model";

export class ContactApi {
  constructor(private http: HttpService) {
  }

  createOrGetContact(createContactParams: CreateContactParams): Promise<ContactId> {
    return this.http.PUT(`/contact`, createContactParams);
  }
  createParentChildrenRelation(request: CreateParentChildrenRequest): Promise<any> {
    return this.http.PUT(`/createParentChildrenRelation`, request);
  }
  getConcessionTypes(): Promise<ConcessionType[]> {
    return this.http.GET(`/getConcessionTypes`);
  }
  getContact(studentUniqueIdentifier: string): Promise<Contact> {
    return this.http.GET(`/contact/${studentUniqueIdentifier}`);
  }
  getContactConcessions(contactIds: string[]): Promise<Concession[]> {
    return this.http.POST(`/getContactConcessions`, contactIds);
  }
  getContactFields(contactFieldsRequest: ContactFieldsRequest): Promise<ContactFields> {
    return this.http.POST(`/contactFields`, contactFieldsRequest);
  }
  submitConcession(concession: Concession): Promise<any> {
    return this.http.PUT(`/submitConcession`, concession);
  }
  submitContactDetails(contactFields: SubmitFieldsRequest): Promise<ContactId> {
    return this.http.POST(`/submitContactDetails`, contactFields);
  }
}

File: willow-api/api/src/main/groovy/ish/oncourse/willow/service/ContactApi.java
package ish.oncourse.willow.service;

import java.util.List;
import ish.oncourse.willow.model.checkout.CreateParentChildrenRequest;
import ish.oncourse.willow.model.checkout.concession.Concession;
import ish.oncourse.willow.model.checkout.concession.ConcessionType;
import ish.oncourse.willow.model.common.CommonError;
import ish.oncourse.willow.model.common.ValidationError;
import ish.oncourse.willow.model.field.ContactFields;
import ish.oncourse.willow.model.field.ContactFieldsRequest;
import ish.oncourse.willow.model.field.SubmitFieldsRequest;
import ish.oncourse.willow.model.web.Contact;
import ish.oncourse.willow.model.web.ContactId;
import ish.oncourse.willow.model.web.CreateContactParams;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.List;
import java.util.Map;
import javax.ws.rs.*;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.MediaType;

@Path("/")
public interface ContactApi  {

    @PUT
    @Path("/contact")
    @Produces({ "application/json" })
    @CollegeInfo
    @ContactValidation
    ContactId createOrGetContact(CreateContactParams createContactParams);

    @PUT
    @Path("/createParentChildrenRelation")
    @Consumes({ "application/json" })
    @Produces({ "application/json" })
    @CollegeInfo
    void createParentChildrenRelation(CreateParentChildrenRequest contactFields);

    @GET
    @Path("/getConcessionTypes")
    @Produces({ "application/json" })
    @CollegeInfo
    List<ConcessionType> getConcessionTypes();

    @GET
    @Path("/contact/{studentUniqueIdentifier}")
    @Produces({ "application/json" })
    @CollegeInfo
    Contact getContact(@PathParam("studentUniqueIdentifier") String studentUniqueIdentifier);

    @POST
    @Path("/getContactConcessions")
    @Produces({ "application/json" })
    @CollegeInfo
    List<Concession> getContactConcessions(List<String> contactIds);

    @POST
    @Path("/contactFields")
    @Consumes({ "application/json" })
    @Produces({ "application/json" })
    @CollegeInfo
    ContactFields getContactFields(ContactFieldsRequest contactFieldsRequest);

    @PUT
    @Path("/submitConcession")
    @Produces({ "application/json" })
    @CollegeInfo
    void submitConcession(Concession concession);

    @POST
    @Path("/submitContactDetails")
    @Consumes({ "application/json" })
    @Produces({ "application/json" })
    @CollegeInfo
    ContactId submitContactDetails(SubmitFieldsRequest contactFields);
}

